samcli/lib/samlib/resource_metadata_normalizer.py

```diff
@@ -144,3 +150,53 @@ def _update_resource_metadata(metadata, updated_values):
         """
         for key, val in updated_values.items():
             metadata[key] = val
+
+    @staticmethod
+    def get_resource_id(resource_properties, logical_id):
+        """
+        Get unique id for a resource.
+        for any resource, the resource id can be the customer defined id if exist, if not exist it can be the
+        cdk-defined resource id, or the logical id if the resource id is not found.
+
+        Parameters
+        ----------
+        resource_properties dict
+            Properties of this resource
+        logical_id str
+            LogicalID of the resource
+
+        Returns
+        -------
+        str
+            The unique function id
+        """
+        resource_metadata = resource_properties.get("Metadata", {})
+        customer_defined_id = resource_metadata.get(SAM_RESOURCE_ID_KEY)
+
+        if isinstance(customer_defined_id, str) and customer_defined_id:
+            return customer_defined_id
+
+        resource_cdk_path = resource_metadata.get(RESOURCE_CDK_PATH_METADATA_KEY)
+
+        if not isinstance(resource_cdk_path, str) or not resource_cdk_path:
+            return logical_id
+
+        # aws:cdk:path metadata format of functions: {stack_id}/{function_id}/Resource
+        # Design doc of CDK path: https://github.com/aws/aws-cdk/blob/master/design/construct-tree.md
+        cdk_path_partitions = resource_cdk_path.split("/")
+
+        if len(cdk_path_partitions) < 2:
+            LOG.warning(
+                "Cannot detect function id from aws:cdk:path metadata '%s', using default logical id", resource_cdk_path
+            )
+            return logical_id
+
+        cdk_resource_id = cdk_path_partitions[-2]
+
+        # Check if the Resource is nested Stack
+        if resource_properties.get("Type", "") == "AWS::CloudFormation::Stack" and cdk_resource_id.endswith(
+            ".NestedStack"
```

Comment:
Please use constant values for some of the string values here (like `AWS::Cloudformation::Stack` or for `.NestedStack`).

Reply:
done